### PR TITLE
Replication namespace

### PR DESCRIPTION
**Pull request by Tokesh**
### Description
adding replication namespace specs

### Issues Resolved
[#232]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
Should we split the docker-compose.yml into docker-compose-default, docker-compose-multinode.yml or something like that?

**Comment by Tokesh** [comment]
> Should we split the docker-compose.yml into docker-compose-default, docker-compose-multinode.yml or something like that?

Oh, do you work even on weekends? 
Yes, I encountered this today, tested it on a local machine. Now I was thinking of asking whether it is worth separating or inventing some flags.

**Comment by dblock** [comment]
Take a look at https://github.com/opensearch-project/opensearch-api-specification/pull/472 that allows custom docker compose and tests.

**Comment by Tokesh** [comment]
> Take a look at #472 that allows custom docker compose and tests.

Thank you! It looks really good
I'll wait for your PR to merge so I can continue working on the replication namespace.

**Comment by dblock** [comment]
> > Take a look at #472 that allows custom docker compose and tests.
> 
> Thank you! It looks really good I'll wait for your PR to merge so I can continue working on the replication namespace.

@Tokesh that got merged FYI

**Comment by Tokesh** [comment]
> > > Take a look at #472 that allows custom docker compose and tests.
> > 
> > 
> > Thank you! It looks really good I'll wait for your PR to merge so I can continue working on the replication namespace.
> 
> @Tokesh that got merged FYI

I have some problems with role mapping. Can you help me please? I got this
(Expected status 200, but received 403: application/json. No user found for indices:admin/plugins/replication/index/setup/validate)

**Comment by dblock** [comment]
Want to finish this @Tokesh?

**Comment by Tokesh** [comment]
> Want to finish this @Tokesh?

Oh, i didnt see your message. Actually, it works on my local computer. I'm still struggling with role mappings

**Comment by dblock** [comment]
> > Want to finish this @Tokesh?
> 
> Oh, i didnt see your message. Actually, it works on my local computer. I'm still struggling with role mappings

LMK what kind of help you need? Rebase and I can help fix any failures here?

**Comment by Tokesh** [comment]
> LMK what kind of help you need? Rebase and I can help fix any failures here?

Yeap, if it's possible please!

When I did everything locally, I did not encounter such problems. Perhaps I missed something somewhere. Perhaps I do not act on the admin side, but I tested different hypotheses. I will be glad if you help me. I tried to send you the documentation, perhaps I read something wrong.

refs:
[*Link*](https://opensearch.org/docs/latest/tuning-your-cluster/replication-plugin/permissions/)

**Comment by dblock** [comment]
@Tokesh For the linter run `npm run lint--fix`, see what comes out?

The errors in the tests seem pretty straightforward:

```
ERROR   RESPONSE STATUS (Expected status 200, but received 403: application/json. No user found for indices:admin/plugins/replication/index/setup/validate)
```

Sounds like something that exists in your instance but not in a clean setup? You can always `docker system prune` and see if you can reproduce?

**Comment by Tokesh** [comment]
> @Tokesh want to finish this?

Hi! I have quite a lot of work at my main job right now. If I can, I'll try to solve this PR on the weekend. If not, then it's worth closing or someone else should finish it if there are people interested. I couldn't do it myself when I was writing tests for CI/CD

**Comment by Tokesh** [comment]
> @Tokesh want to finish this?

I passed the biggest part of CI/CD. 
Any suggestions to refactor this code?

**Review by dblock — DISMISSED** [review]
This looks great!

I added .yml files to Vale preprocessing in https://github.com/opensearch-project/opensearch-api-specification/pull/660, let's rebase this once it's merged to make sure it worked.

File: .github/workflows/test-spec.yml
Comment (dblock) [inline .github/workflows/test-spec.yml:82]: Get rid of the extra spaces.

File: tests/plugins/replication/replica.yaml
Comment (Tokesh) [reply tests/plugins/replication/replica.yaml:1]: I think i can divide to:
```replica.yaml```
```stats.yaml```
```autofollow.yaml```
```replica-rules.yaml```
And the code will be duplicated everywhere where we create an index and start replicating it, if it works, then I can divide it like this and do it in a few days.

Comment (dblock) [reply tests/plugins/replication/replica.yaml:1]: That works. Maybe rename replica.yaml to index.yaml, that one will have the replication as the main act.

**Comment by Tokesh** [comment]
Ready for your review! @dblock @nhtruong

File: .github/workflows/test-spec.yml
Comment (dblock) [inline .github/workflows/test-spec.yml:39]:
```suggestion
          - version: 2.18.0
```

Let's just use the latest, update 2.16 here to 2.18 and remove 2.17.

File: spec/namespaces/replication.yaml
Comment (dblock) [inline spec/namespaces/replication.yaml:254]: Use a `$ref` to `AcknowledgedResponseBase`,  https://github.com/opensearch-project/opensearch-api-specification/blob/bf356019ed3450d97fb4f9f68be4db86d1ed2195/spec/schemas/_common.yaml#L14

File: spec/schemas/replication._common.yaml
Comment (dblock) [inline spec/schemas/replication._common.yaml:29]: Is this a known enum? If so it should be one.

File: spec/schemas/replication._common.yaml
Comment (dblock) [inline spec/schemas/replication._common.yaml:35]: If this is an index name it should be a ref.

File: spec/schemas/replication._common.yaml
Comment (dblock) [inline spec/schemas/replication._common.yaml:48]: Should be 

```
          $ref: '_common.yaml#/components/schemas/SequenceNumber'
```

File: spec/schemas/replication._common.yaml
Comment (dblock) [inline spec/schemas/replication._common.yaml:62]: Time values in ms are typically 

```
$ref: '_common.yaml#/components/schemas/StringifiedEpochTimeUnitMillis'
```

File: spec/schemas/replication._common.yaml
Comment (dblock) [inline spec/schemas/replication._common.yaml:56]: Use a ref.

**Review by dblock — CHANGES_REQUESTED** [review]
Glad we have a passing build!

Go through the fields and check whether they are semantically meaningful and need to be a `$ref` to something that exists (time values, index names, etc.).

File: spec/schemas/replication._common.yaml
Comment (dblock) [inline spec/schemas/replication._common.yaml:198]: Is this a combination of `SettingsBody` and these? So should be `allOf`.

**Review by dblock — CHANGES_REQUESTED** [review]
Small one left, the rest LGTM!

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
I made the change I suggested above and merged, thank you, this is great work @Tokesh!

**Comment by Tokesh** [comment]
Oh, i just checked. Thank you so much for detailed reviews and help! It's a pleasure working with you!

**Pull request merged**
